cluster/types.go

```diff
@@ -0,0 +1,212 @@
+//                              │    ╠══════════════════╗       │         │
+//                              │    ║                  ║       │         │
+//                   Teardown() │    ║      Setup       ║◀────Reset()     │
+//                      ┌───────┘    ║                  ║        │        │
+//                      │       ┌───▶╚═══╦══════════════╝        │        │
+//                      │       │        │     │                 │        │
+//                      │     Reset()────┘     │ Initialize()    │        │
+//                      │       │              └────────┐        │        │
+//                      │       │                       │        │        │
+//                      │       │                       │        │        │
+//                      │       │                       ▼        │        │
+//            ╔═════════╩═══════╩╗   Start()           ╔═════════╩════════╣
+//            ║                  ║   StartInitialized()║                  ║
+//         ┌──║     Running      ║◀────────────────────║   Initialized    ║──┐
+//         │  ║                  ║                     ║                  ║  │
+//         │  ╚══════════════════╩───────Stop()───────▶╚══════════════════╝  │
+// AddInstance()        ▲                                        ▲           │
+// RemoveInstance()     │                                        │     AddInstance()
+// ReplaceInstance()────┘                                        └─────RemoveInstance()
+//                                                                     ReplaceInstance()
+type Status int
+
+const (
+	// ClusterStatusUninitialized refers to the state of an un-initialized cluster.
+	ClusterStatusUninitialized Status = iota
+
+	// ClusterStatusSetup refers to the state of a cluster whose instances have been
+	// setup. In this state, the instances are not running, and the cluster services
+	// do not have a defined placement.
+	ClusterStatusSetup
+
+	// ClusterStatusInitialized refers to the state of a cluster whose instances have
+	// been setup, and the cluster has an assigned placement.
+	ClusterStatusInitialized
+
+	// ClusterStatusRunning refers to the state of a cluster with running instances.
+	// There is no restriction on the number of instances running, or assigned within
+	// the cluster placement.
+	ClusterStatusRunning
+
+	// ClusterStatusError refers to a cluster in error.
+	ClusterStatusError
+)
+
+// Cluster is a collection of clustered M3DB instances.
+type Cluster interface {
+	// Setup the instances in the Environment provided during construction.
+	Setup() error
+
+	// Initialize initializes service placement for the specified numNodes.
+	Initialize(numNodes int) ([]env.M3DBInstance, error)
+
+	// AddInstance adds the specified instance to the service placement. It does
+	// NOT alter the state of the ServiceInstance (i.e. does not start/stop it).
+	AddInstance() (env.M3DBInstance, error)
```

Comment:
could modify the placementservice interface to return the instance used in the AddInstance() operation: https://github.com/m3db/m3cluster/blob/master/services/types.go#L184